### PR TITLE
[JENKINS-65611] Ensure facets are modifiable

### PR DESCRIPTION
**Pull request by res0nance**
See [JENKINS-65611](https://issues.jenkins-ci.org/browse/JENKINS-65611).

### Proposed changelog entries

* Entry 1: JENKINS-65611, Fix a regression where Fingerprint Facets were not removable
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@basil 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
We probably want to ship a new weekly with this in it

File: test/src/test/java/hudson/model/FingerprintTest.java [hunk 1]
```diff
@@ -413,6 +414,24 @@ public void checkNormalFingerprintWithWebClient() throws Exception {
...
+        assertThat(fingerprint.getFacets().size(), is(0));
+        fingerprint.getFacets().add(testFacet);
+        assertThat(fingerprint.getFacets().size(), is(1));
+        assertTrue(fingerprint.getFacets().contains(testFacet));
+        fingerprint.getFacets().remove(testFacet);
+        assertThat(fingerprint.getFacets().size(), is(0));
+        fingerprint.getFacets().add(testFacet);
+        Iterator<FingerprintFacet> itr = fingerprint.getFacets().iterator();
+        itr.next();
+        itr.remove();
```
Comment (jtnord) [inline test/src/test/java/hudson/model/FingerprintTest.java:432]: I think you want to actually do a test here to check the thing is actually removed as opposed to silently fails.

Looking at the code I am not convinced that this is anything other than a no-op

https://github.com/jenkinsci/jenkins/blob/a1175b932e1ac97085d69752f5821141d012314f/core/src/main/java/hudson/util/Iterators.java#L337-L344

**Review by jtnord — CHANGES_REQUESTED** [review]

**Comment by MarkEWaite** [comment]
> We probably want to ship a new weekly with this in it

I agree that we need a new weekly that includes a fix for [JENKINS-65611](https://issues.jenkins.io/browse/JENKINS-65611)

File: test/src/test/java/hudson/model/FingerprintTest.java (hunk 1 above)
Comment (res0nance) [reply test/src/test/java/hudson/model/FingerprintTest.java:432]: That function is a little misleading it is actually https://github.com/jenkinsci/jenkins/blob/a1175b932e1ac97085d69752f5821141d012314f/core/src/main/java/hudson/util/Iterators.java#L367-L369

**Review by jtnord — APPROVED** [review]

**Comment by timja** [comment]
I propose we ship a release after the tests pass and merge this, any objections @jenkinsci/core?

**Comment by timja** [comment]
I propose we ship a release after the tests pass and merge this, any objections @jenkinsci/core?

**Review by oleg-nenashev — APPROVED** [review]
+1 for out of order release. Time to improve test coverage

**Review by basil — APPROVED** [review]
I would like to apologize for the inconvenience caused by this bug and to thank everyone who spent time debugging and fixing it.

I think this bug could have been avoided if I had looked at the callers of `Fingerprint#getFacets`. Then I would have seen that it has no test coverage, and that would have prompted me to look at external usages with `usages-in-plugins` and test them or to add an integration test as Raihaan has done here. I will do that in the future.

**Comment by res0nance** [comment]
> I propose we ship a release after the tests pass and merge this, any objections @jenkinsci/core?

Can we somehow stop distribution of 2.292 as well? This will prevent this bug affecting more users

**Comment by MarkEWaite** [comment]
> Can we somehow stop distribution of 2.292 as well? This will prevent this bug affecting more users

I'd rather spend the effort delivering the 2.293 release out of band today rather than spend the time to stop distribution of 2.292 and then spend the time to create the 2.293 out of band release as well.

**Comment by jglick** [comment]
> if I had looked at the callers…

Well the better defense would be to run `plugin-compat-tester` on at least some common plugins before shipping a core release, since from the look of the stack trace it seems pretty likely this would have shown up in multiple test regressions.

**Comment by MarkEWaite** [comment]
Build is running for a new weekly release, Jenkins 2.293.  Thanks @res0nance and @timja !

**Comment by MarkEWaite** [comment]
Build has completed.  I've confirmed that the Docker image, the war file, the deb, and the Red Hat rpm installed successfully.

**Pull request merged**